src/NovelRT/Experimental/Graphics/Vulkan/VulkanGraphicsContext.cpp

```diff
@@ -0,0 +1,519 @@
+        imageSubresourceLayers.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT;
+        imageSubresourceLayers.layerCount = 1;
+
+        VkExtent3D imageExtent{};
+        imageExtent.width = destination->GetWidth();
+        imageExtent.height = destination->GetHeight();
+        imageExtent.depth = destination->GetDepth();
+
+        VkBufferImageCopy vulkanBufferImageCopy{};
+        vulkanBufferImageCopy.imageSubresource = imageSubresourceLayers;
+        vulkanBufferImageCopy.imageExtent = imageExtent;
+
+        vkCmdCopyBufferToImage(vulkanCommandBuffer, source->GetVulkanBuffer(), vulkanImage,
+                               VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL, 1, &vulkanBufferImageCopy);
+
+        EndCopy(vulkanImage);
+    }
+
+    void VulkanGraphicsContext::Draw(const std::shared_ptr<VulkanGraphicsPrimitive>& primitive)
+    {
+        if (primitive == nullptr)
+        {
+            throw Exceptions::NullPointerException("The provided GraphicsPrimitive is null.");
+        }
+
+        VkCommandBuffer vulkanCommandBuffer = GetVulkanCommandBuffer();
+        std::shared_ptr<VulkanGraphicsPipeline> pipeline = primitive->GetPipeline();
+        std::shared_ptr<VulkanGraphicsPipelineSignature> pipelineSignature = pipeline->GetSignature();
+        VkPipeline vulkanPipeline = pipeline->GetVulkanPipeline();
+
+        vkCmdBindPipeline(vulkanCommandBuffer, VK_PIPELINE_BIND_POINT_GRAPHICS, vulkanPipeline);
+
+        const GraphicsMemoryRegion<GraphicsResource>& vertexBufferRegion = primitive->GetVertexBufferRegion();
+        std::shared_ptr<VulkanGraphicsBuffer> vertexBuffer =
+            std::static_pointer_cast<VulkanGraphicsBuffer>(vertexBufferRegion.GetCollection());
+        VkBuffer vulkanVertexBuffer = vertexBuffer->GetVulkanBuffer();
+        size_t vulkanVertexBufferOffset = vertexBufferRegion.GetOffset();
+
+        vkCmdBindVertexBuffers(vulkanCommandBuffer, 0, 1, &vulkanVertexBuffer, &vulkanVertexBufferOffset);
+
+        VkDescriptorSet vulkanDescriptorSet = pipelineSignature->GetVulkanDescriptorSet();
+
+        if (vulkanDescriptorSet != VK_NULL_HANDLE)
+        {
+            gsl::span<const GraphicsMemoryRegion<GraphicsResource>> inputResourceRegions =
+                primitive->GetInputResourceRegions();
+            size_t inputResourceRegionsLength = inputResourceRegions.size();
+
+            for (size_t index = 0; index < inputResourceRegionsLength; index++)
+            {
+                const GraphicsMemoryRegion<GraphicsResource>& inputResourceRegion = inputResourceRegions[index];
+
+                VkWriteDescriptorSet writeDescriptorSet{};
+
+                std::shared_ptr<VulkanGraphicsBuffer> vulkanGraphicsBuffer =
+                    std::dynamic_pointer_cast<VulkanGraphicsBuffer>(inputResourceRegion.GetCollection());
+                std::shared_ptr<VulkanGraphicsTexture> vulkanGraphicsTexture =
+                    std::dynamic_pointer_cast<VulkanGraphicsTexture>(inputResourceRegion.GetCollection());
```

Comment:
I know the point of the checks here is to see what kind of object the `InputResourceRegion`'s collection is - but is there no better way to do this? I'm not sure that desiring for one of these checks to return a `nullptr` is something we explicitly _want_... I get if there's no real better way to do this though.

Reply:
This is how it worked in the original implementation, but if you can offer a better design suggestion I'm all ears. The only thing I can think of is an explicit parameter to tell it what type it is, but that sounds unintuitive from an API standpoint.